### PR TITLE
Neo4j requests

### PR DESCRIPTION
**Pull request by scferrada**
I just left the neo4j responses as Strings, I might add a more suitable POJO class to parse such responses if needed.

### CODE REVIEW

File: src/main/java/se/liu/ida/hefquin/engine/federation/access/Neo4JConnectionException.java [hunk 1]
```diff
@@ -0,0 +1,6 @@
+package se.liu.ida.hefquin.engine.federation.access;
+
+public class Neo4JConnectionException extends RuntimeException {
+    public Neo4JConnectionException(String s) {
```
Comment (hartig) [inline src/main/java/se/liu/ida/hefquin/engine/federation/access/Neo4JConnectionException.java:4]:
```suggestion
    public Neo4JConnectionException( final String msg ) {
        super(msg);
```

File: src/main/java/se/liu/ida/hefquin/engine/federation/access/Neo4jConnectionFactory.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+public class Neo4jConnectionFactory {
+
+    static public Neo4jConnection connect(final String URL){
+        return new Neo4jConnection(URL);
+    }
+
+    public static class Neo4jConnection {
+        private String URL;
+
+        public Neo4jConnection(String URL) {
```
Comment (hartig) [inline src/main/java/se/liu/ida/hefquin/engine/federation/access/Neo4jConnectionFactory.java:18]:
```suggestion
        public Neo4jConnection( final String URL ) {
```

File: src/main/java/se/liu/ida/hefquin/engine/federation/access/Neo4jConnectionFactory.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+import java.net.http.HttpResponse;
+
+public class Neo4jConnectionFactory {
+
+    static public Neo4jConnection connect(final String URL){
+        return new Neo4jConnection(URL);
+    }
+
+    public static class Neo4jConnection {
+        private String URL;
```
Comment (hartig) [inline src/main/java/se/liu/ida/hefquin/engine/federation/access/Neo4jConnectionFactory.java:16]:
```suggestion
        protected final String URL;
```

File: src/main/java/se/liu/ida/hefquin/engine/federation/access/Neo4jConnectionFactory.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+                    .POST(HttpRequest.BodyPublishers.ofString(data))
+                    .build();
+
+            HttpClient client = HttpClient.newHttpClient();
+            try {
+                HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
+                return response.body();
+            } catch (IOException e) {
+                throw new Neo4JConnectionException("Data could not be sent to the server");
+            } catch (InterruptedException e) {
```
Comment (hartig) [inline src/main/java/se/liu/ida/hefquin/engine/federation/access/Neo4jConnectionFactory.java:44]:
```suggestion
        public String executeQuery( final String cypher ) {
            final String data = "{\n" +
                    "  \"statements\" : [ {\n" +
                    "    \"statement\" : \""+cypher+"\",\n" +
                    "    \"parameters\" : {}\n" +
                    "  } ]\n" +
                    "}";

            final var request = HttpRequest.newBuilder(
                    URI.create(this.URL))
                    .header("Accept", "application/json;charset=UTF-8")
                    .header("Content-Type", "application/json")
                    .POST(HttpRequest.BodyPublishers.ofString(data))
                    .build();

            final HttpClient client = HttpClient.newHttpClient();
            try {
                final HttpResponse<String> response = client.send(request, HttpResponse.BodyHandlers.ofString());
                return response.body();
            } catch ( final IOException e ) {
                throw new Neo4JConnectionException("Data could not be sent to the server");
            } catch ( final InterruptedException e ) {
```

File: src/main/java/se/liu/ida/hefquin/engine/federation/access/Neo4jInterface.java [hunk 5]
```diff
@@ -0,0 +1,6 @@
+package se.liu.ida.hefquin.engine.federation.access;
+
+public interface Neo4jInterface extends SolMapRetrievalInterface {
```
Comment (hartig) [inline src/main/java/se/liu/ida/hefquin/engine/federation/access/Neo4jInterface.java:3]: The way it is currently implemented, `Neo4jInterface` is not a `SolMapRetrievalInterface`. I suggest to introduce a new interface called `StringRetrievalInterface` and make `Neo4jInterface` an extension of that one.

File: src/main/java/se/liu/ida/hefquin/engine/federation/access/impl/BlockingFederationAccessManagerImpl.java [hunk 6]
```diff
@@ -46,4 +38,9 @@ public TPFResponse performRequest( final BRTPFRequest req, final BRTPFServer fm)
 		return reqProcBRTPF.performRequest( req, fm );
 	}
 
+	@Override
+	public StringRetrievalResponse performRequest(Neo4jRequest req, Neo4jServer fm) {
```
Comment (hartig) [inline src/main/java/se/liu/ida/hefquin/engine/federation/access/impl/BlockingFederationAccessManagerImpl.java:42]:
```suggestion
	public StringRetrievalResponse performRequest( final Neo4jRequest req, final Neo4jServer fm ) {
```

File: src/main/java/se/liu/ida/hefquin/engine/federation/access/impl/FederationAccessManagerBase.java [hunk 7]
```diff
@@ -15,18 +12,23 @@
...
 	protected final TPFRequestProcessor       reqProcTPF;
 	protected final BRTPFRequestProcessor     reqProcBRTPF;
+	protected final Neo4jRequestProcessor	  reqProcNeo4j;
 
 	protected FederationAccessManagerBase(
 			final SPARQLRequestProcessor reqProcSPARQL,
 			final TPFRequestProcessor reqProcTPF,
-			final BRTPFRequestProcessor reqProcBRTPF ) {
+			final BRTPFRequestProcessor reqProcBRTPF,
+			Neo4jRequestProcessor reqProcNeo4j) {
```
Comment (hartig) [inline src/main/java/se/liu/ida/hefquin/engine/federation/access/impl/FederationAccessManagerBase.java:21]:
```suggestion
			final Neo4jRequestProcessor reqProcNeo4j ) {
```

File: src/main/java/se/liu/ida/hefquin/engine/federation/access/impl/iface/Neo4jInterfaceImpl.java [hunk 8]
```diff
@@ -0,0 +1,35 @@
+package se.liu.ida.hefquin.engine.federation.access.impl.iface;
+
+import se.liu.ida.hefquin.engine.federation.access.DataRetrievalRequest;
+import se.liu.ida.hefquin.engine.federation.access.Neo4jInterface;
+import se.liu.ida.hefquin.engine.federation.access.Neo4jRequest;
+
+public class Neo4jInterfaceImpl implements Neo4jInterface {
+
+    final String url;
```
Comment (hartig) [inline src/main/java/se/liu/ida/hefquin/engine/federation/access/impl/iface/Neo4jInterfaceImpl.java:9]:
```suggestion
    protected final String url;
```

File: src/main/java/se/liu/ida/hefquin/engine/federation/access/impl/iface/Neo4jInterfaceImpl.java [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+        return false;
+    }
+
+    @Override
+    public boolean supportsBGPRequests() {
+        return false;
+    }
+
+    @Override
+    public boolean supportsRequest(DataRetrievalRequest req) {
```
Comment (hartig) [inline src/main/java/se/liu/ida/hefquin/engine/federation/access/impl/iface/Neo4jInterfaceImpl.java:27]:
```suggestion
    public boolean supportsRequest( final DataRetrievalRequest req ) {
```

File: src/main/java/se/liu/ida/hefquin/engine/federation/access/impl/req/Neo4jRequestImpl.java [hunk 10]
```diff
@@ -0,0 +1,24 @@
+package se.liu.ida.hefquin.engine.federation.access.impl.req;
+
+import se.liu.ida.hefquin.engine.federation.access.Neo4jRequest;
+import se.liu.ida.hefquin.engine.queryplan.ExpectedVariables;
+
+public class Neo4jRequestImpl implements Neo4jRequest {
+
+    final private String cypherQuery;
```
Comment (hartig) [inline src/main/java/se/liu/ida/hefquin/engine/federation/access/impl/req/Neo4jRequestImpl.java:8]:
```suggestion
    protected final String cypherQuery;
```

File: src/main/java/se/liu/ida/hefquin/engine/federation/access/impl/reqproc/Neo4jRequestProcessorImpl.java [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+import se.liu.ida.hefquin.engine.federation.access.impl.response.StringResponseImpl;
+
+import java.util.Date;
+
+public class Neo4jRequestProcessorImpl implements Neo4jRequestProcessor{
+    @Override
+    public StringRetrievalResponse performRequest(Neo4jRequest req, Neo4jServer fm) {
+        final Neo4jConnectionFactory.Neo4jConnection conn = Neo4jConnectionFactory.connect(fm.getInterface().getURL());
+        final Date requestStartTime = new Date();
+        String result = conn.executeQuery(req.getCypherQuery());
```
Comment (hartig) [inline src/main/java/se/liu/ida/hefquin/engine/federation/access/impl/reqproc/Neo4jRequestProcessorImpl.java:14]:
```suggestion
    public StringRetrievalResponse performRequest( final Neo4jRequest req, final Neo4jServer fm ) {
        final Date requestStartTime = new Date();
        final Neo4jConnectionFactory.Neo4jConnection conn = Neo4jConnectionFactory.connect( fm.getInterface().getURL() );
        final String result = conn.executeQuery(req.getCypherQuery());
```

Comment (hartig) [reply src/main/java/se/liu/ida/hefquin/engine/federation/access/impl/reqproc/Neo4jRequestProcessorImpl.java:14]: Another thought here, not for this PR but as a more general thing for the future: We should avoid having to open a new connection for every request. Hence, we need something like a connection manager from which a required connection can be obtained and that takes care of opening and closing the connections on demand.

File: src/test/java/se/liu/ida/hefquin/engine/EngineTestBase.java [hunk 12]
```diff
@@ -330,6 +316,11 @@ public TPFResponse performRequest( final BRTPFRequest req, final BRTPFServer fm
 				return ( (BRTPFServerForTest) fm ).performRequest(req);
 			}
 		}
+
+		@Override
+		public StringRetrievalResponse performRequest(Neo4jRequest req, Neo4jServer fm) {
```
Comment (hartig) [inline src/test/java/se/liu/ida/hefquin/engine/EngineTestBase.java:321]:
```suggestion
		public StringRetrievalResponse performRequest( final Neo4jRequest req, final Neo4jServer fm ) {
```

File: src/test/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/pullbased/TestUtils.java [hunk 13]
```diff
@@ -98,5 +91,10 @@ public TPFResponse performRequest( final TPFRequest req, final BRTPFServer fm )
 		public TPFResponse performRequest( final BRTPFRequest req, final BRTPFServer fm ) {
 			return null;
 		}
+
+		@Override
+		public StringRetrievalResponse performRequest(Neo4jRequest req, Neo4jServer fm) {
```
Comment (hartig) [inline src/test/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/pullbased/TestUtils.java:96]:
```suggestion
		public StringRetrievalResponse performRequest( final Neo4jRequest req, final Neo4jServer fm ) {
```

**Review by hartig — CHANGES_REQUESTED** [review]
Looks good in general!! I have just several minor edits. Except for one, you can simply accept all of them within the github Web app.

**Comment by hartig** [comment]
I see that you have add some changes. Let me know when you are done. Then, I can merge this PR.

**Comment by scferrada** [comment]
I have just pushed the last changes. Thanks!

**Comment by scferrada** [comment]
Oh, it seems there are some conflicts, I will rebase the main branch and solve them.

**Review by hartig — APPROVED** [review]
All good now. I will take care of merging this PR.

**Comment by hartig** [comment]
No need to rebase. I have already resolved the conflicts manually.

**Pull request merged**